### ISSUE TITLE
Compiler error when using "Serializing with source generation"

### ISSUE THREAD

**Issue by Steve-OH** [SteveDunn/Vogen#622]
### Describe the bug

IDE: Visual Studio 2022
Build target: .NET 8
Language: C# 12
Vogen: 4.0.8

Attempting to follow the example at https://stevedunn.github.io/Vogen/using-with-json.html#serializing-with-source-generation results in a compiler error in the generated code: CS0034: Operator '==' is ambiguous on operands of type 'Foo' and '&lt;null&gt;'.

### Steps to reproduce

Attempt to compile this simple console app:

```csharp
namespace MyNamespace;

using System.Text.Json.Serialization;
using Vogen;

[ValueObject<string>]
internal partial record Foo;

[JsonSourceGenerationOptions(WriteIndented = true)]
[JsonSerializable(typeof(Foo))]
internal partial class JsonSourceGenerationContext :
    JsonSerializerContext;

internal class Program
{
    static void Main()
    {
    }
}
```

### Expected behaviour

There should not be a compiler error.

**Comment by SteveDunn** [comment]
Hi @Steve-OH - thank you for the bug report.  What's causing that (which isn't the underlying issue), is because Vogen generates two `operator ==` (one for the primitive, one for the wrapper). Hence, the comparison is ambiguous. STJ should generate `is null` rather than `== null`, and that's a bug I plan to raise and fix shortly.

The underlying issue is that source generators cannot rely on other source generators. That is, when we create the JSON context, we say that there's a type converter for `Foo`, but the type converter doesn't exist yet (just the `partial record Foo` in your user code). So STJ generates its own, and it is in this generated code that the ambiguous `operator ==` is generated.

So, to use source-generated converters in source-generated STJ scenarios, we need to use the generated `type factory` which is emitted by Vogen during the build. At runtime, you tell STJ about it:

```c#
Person person = new()
{
    Name = Name.From("Fred Flintstone"),
    Age = Age.From(44),
    Address = Address.From("201 Cobblestone Lane"),
};

var options = new JsonSerializerOptions
{
    WriteIndented = true,
    Converters =
    {
        new VogenTypesFactory()
    }
};

var ctx = new JsonSourceGenerationContext(options);

var json = JsonSerializer.Serialize(person, ctx.Person);
Person person2 = JsonSerializer.Deserialize(json, ctx.Person)!;
```

The `VogenTypesFactory` is a generated lookup of type to converter, e.g.

```c#
public class VogenTypesFactory : JsonConverterFactory
{
    private static readonly Dictionary<Type, Lazy<JsonConverter>> _lookup = 
        new()
        {
            { typeof(Age), new Lazy<JsonConverter>(() => new Age.AgeSystemTextJsonConverter()) },
            { typeof(Name), new Lazy<JsonConverter>(() => new Name.NameSystemTextJsonConverter()) },
            { typeof(Address), new Lazy<JsonConverter>(() => new Address.AddressSystemTextJsonConverter()) }
        };
    
    public override bool CanConvert(Type typeToConvert) => _lookup.ContainsKey(typeToConvert);
    
    public override JsonConverter CreateConverter(Type typeToConvert, System.Text.Json.JsonSerializerOptions options) =>
        _lookup[typeToConvert].Value;
}
```

I hope that helps. I'll update the documentation to make it clearer how to use this.

To get your code working, you can change your value object to be a `struct` rather than a `record class`. In fact, [it is recommended ](https://stevedunn.github.io/Vogen/records.html) to *not* use records as the additional functionality that is generated by the compiler offers no value.

It is also worth noting that if you value objects were defined in another assembly, then the problem of source generators using source generators is not a problem, as, by the time you reference the value object from your other project, it is fully formed. Bbut sadly, there's a bug in Vogen that means that the STJ converter is `internal`. But happily, that's now fixed and will be in the next release, hopefully later today.

Thanks again for the bug report. I'll close this for now, but feel free to request a reopening if what I've covered isn't applicable to your particular scenario.

**Comment by Steve-OH** [comment]
I don't actually need to use this functionality. The issue arose while I was investigating a different problem (#624), and I just noticed the bug in transit, so to speak, so that's why I reported it. I've since figured out what was causing my original problem, which is what #624 is all about.

There **_is_** a reason to use `record` in some cases: you want the record semantics, and you also want inheritance. I have some closely-related ID value objects that share various constraints, so it makes sense to derive them from an abstract base record that implements those constraints.
